File: yologram-api-v1/src/main/kotlin/link/yologram/api/v1/infra/client/cms/CmsApiClient.kt
package link.yologram.api.v1.infra.client.cms

/**
 * cms 도메인 경계 호출 추상화 (카테고리 활성 검증) — pms 등 소비 도메인이 사용.
 * 모놀리식에서는 cms 리포지토리를 직접 호출(LocalCmsApiClient)하고,
 * MSA 분리 시 이 패키지에 RestCmsApiClient + Config + dto를 추가해 교체한다
 * (번장 bun-order-api의 infra/noti/client 구성 미러).
 */
interface CmsApiClient {
    /** categoryIds가 모두 테크 게시판의 활성 카테고리이면 true. 빈 목록은 true. */
    fun allActive(categoryIds: Collection<Long>): Boolean

    /** categoryId → 카테고리명 배치 조회 (라벨 해석). 존재하지 않는(삭제된) id는 결과에서 제외. */
    fun findCategoryNames(categoryIds: Collection<Long>): Map<Long, String>
}

File: yologram-api-v2/app/infra/client/cms/cms_api_client.py
# 타 도메인 경계 클라이언트 — 타 도메인 리포지토리 import는 app/infra/client에서만 허용 (domain 간 직접 참조 금지).
# MSA 분리 시 Local 구현 대신 Rest(HTTP) 구현을 추가해 교체한다.
from collections.abc import Collection
from typing import Protocol

from sqlalchemy.orm import Session

from app.domain.cms.tech.repository import TechCategoryRepository


class CmsApiClient(Protocol):
    """
    타 도메인(pms·news) → tech category 도메인 경계 호출 추상화.
    모놀리식에서는 category 리포지토리를 직접 호출(LocalCmsApiClient),
    MSA 분리 시 category-api HTTP 호출 구현으로 교체한다.
    """

    def all_active(self, category_ids: Collection[int]) -> bool:
        """category_ids가 모두 테크 게시판의 활성 카테고리이면 True. 빈 목록은 True."""
        ...

    def find_category_names(self, category_ids: Collection[int]) -> dict[int, str]:
        """카테고리 라벨 일괄 조회 (N+1 회피). id→name, 삭제된 id는 결과에서 제외."""
        ...


class LocalCmsApiClient:

    def __init__(self, db: Session):
        self.repository = TechCategoryRepository(db)

    def all_active(self, category_ids: Collection[int]) -> bool:
        distinct_ids = list(set(category_ids))
        if not distinct_ids:
            return True
        matched = self.repository.count_active_by_ids(distinct_ids)
        return matched == len(distinct_ids)

    def find_category_names(self, category_ids: Collection[int]) -> dict[int, str]:
        distinct_ids = list(set(category_ids))
        if not distinct_ids:
            return {}
        # 라벨 해석은 is_active 무관 (api-v1 findAllById 정합) — 삭제된 id는 조회 결과에 없어 자연 제외
        return {c.id: c.name for c in self.repository.find_by_ids(distinct_ids)}
